### DISCUSSION TITLE
Where are the boundaries for client components when using context providers?

### DISCUSSION THREAD

**Discussion by 0xabrar** [vercel/next.js#51137] [Help]
### Summary

I was looking over what's required to move from from the Page router over to the App router, and one of the things that came up is how to handle ContextProviders that are included at the root of our application. The docs suggest the following: https://nextjs.org/docs/getting-started/react-essentials#rendering-third-party-context-providers-in-server-components

I'm a bit confused on where the cut-off between client and server components happens then. Earlier in the doc (https://nextjs.org/docs/getting-started/react-essentials#the-use-client-directive) it's mentioned that 'use client' marks the boundary between the server component graph and client component module graph. So if I have use-client in a provider at the root of my app, doesn't that just make the entire thing non-RSCs? Or did I misunderstand here.

Also if all of the child components (the whole app) then becomes client components, what are my alternatives here? There's a bunch of 3rd party packages + auth in my app that are at the root through the context API.

Update:
This section of the docs actually clears it up that when you pass children as props these are "holes" in the client component which can take server components. https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props

Guess that clears it up? Seems like you can have server components past the boundary of "use client".

**Comment by devjiwonchoi** [comment]
This was one of my troubleshoots, and am glad to share it.

## Where does the line drawn between the Client & Server Components?

To keep it concise, Client Components are the **specific file**, which has been declared `'use client'` on the top. Server Components are the rest that is outside of those cases.

> Since Server Components are the default, all components are part of the Server Component module graph unless defined or imported in a module that starts with the "use client" directive.

## So the component inside the Client Components are considered as Client Module Graph?

Technically, yes.

> Once "use client" is defined in a file, all other modules imported into it, including child components, are considered part of the client bundle.

Are they considered Client Components? NO. 
When rendering the Server & Client Components, the Server Components within the client Components are prioritized to be rendered.

> On the server, React renders all Server Components before sending the result to the client.
This includes Server Components nested inside Client Components.

## What happens if "use client" on the VERY Root Provider?

This is the tricky part since the Server Components nested inside the Client Components ARE part of the Client Module Graph.
Referring to the [docs](https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props), the Server Components nested as `children` will be rendered on the server but included in the Client Module Graph as a "hole", and will be "filled" when the Client is rendered.

> The Server Component will be rendered on the server, and when the Client Component is rendered on the client, the "hole" will be filled in with the rendered result of the Server Component.

Therefore, the Server Component `children` of the Client Component Root Provider will be server-rendered.

## Then is it recommended to `use client` the Root Layout?

The answer is no. Although the Server Components will be server-rendered, they will still be part of the Client Module Graph, which will consume the bundle size. Therefore the [docs](https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves) are recommending moving Client Components as "Leaves" as possible.

> To improve the performance of your application, we recommend moving Client Components to the leaves of your component tree where possible.

## Conclusion

Server Components nested in Client Components as `children` are considered client bundles and will partially consume size, but will be server-rendered and behave as a hole (similar to a Promise I would say) that will be filled when the corresponding Client Component is rendered.

P.S. Please feel free to give feedback. 😉

## Reference
- [Client Components](https://nextjs.org/docs/getting-started/react-essentials#client-components)
- [Moving Client Components to the Leaves](https://nextjs.org/docs/getting-started/react-essentials#moving-client-components-to-the-leaves)
- [Composing Client and Server Components](https://nextjs.org/docs/getting-started/react-essentials#composing-client-and-server-components)
- [Passing Server Components to Client Components as Props](https://nextjs.org/docs/getting-started/react-essentials#recommended-pattern-passing-server-components-to-client-components-as-props)
